DPL Analysis: introduce PersistentColumn concept

ktf opened the pull request:
DPL Analysis: introduce PersistentColumn concept

This way the compiler does not need to instanciate the conjuction type
and it will point to exactly which column is not persistent, when needed.

alibuild commented:
Error while checking build/O2/fullCI for 29e513b5f284a9986d6fab4ed23618fc4200bb55 at 2024-02-02 15:19:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
/sw/SOURCES/O2/12648-slc8_x86-64/0/Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h:246:15: error: field declaration 'cSpeed' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:29:7: error: field reference 'nChanA' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:29:7: error: field reference 'nChanC' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:52:5: error: field reference 'nChanA' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:53:5: error: field reference 'nChanC' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:73:38: error: field reference 'nChanA' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:74:38: error: field reference 'nChanC' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:83:5: error: field reference 'nChanA' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:84:5: error: field reference 'nChanC' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:101:5: error: field reference 'nChanA' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:101:14: error: field reference 'nChanC' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:108:37: error: field reference 'nChanA' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:108:45: error: field reference 'nChanC' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:109:49: error: field reference 'nChanA' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:109:63: error: field reference 'nChanC' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:118:11: error: field declaration 'nChanA' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/DataFormats/Detectors/FIT/common/include/DataFormatsFIT/Triggers.h:119:11: error: field declaration 'nChanC' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h:246:15: error: field declaration 'cSpeed' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/Detectors/AOD/src/AODProducerWorkflowSpec.cxx:2392:72: error: field reference 'cSpeed' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/Detectors/AOD/src/AODProducerWorkflowSpec.cxx:2401:59: error: field reference 'cSpeed' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/Detectors/AOD/src/AODProducerWorkflowSpec.cxx:2754:74: error: field reference 'cSpeed' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/Detectors/AOD/src/AODProducerWorkflowSpec.cxx:2764:39: error: field reference 'cSpeed' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:146:57: error: field reference 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:206:62: error: field reference 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:277:9: error: field declaration 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:322:50: error: field reference 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:346:19: error: field reference 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:347:19: error: field reference 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:146:57: error: field reference 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:206:62: error: field reference 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:277:9: error: field declaration 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:322:50: error: field reference 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:346:19: error: field reference 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:347:19: error: field reference 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:146:57: error: field reference 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:206:62: error: field reference 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:277:9: error: field declaration 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
/sw/SOURCES/O2/12648-slc8_x86-64/0/GPU/TPCFastTransformation/TPCFastSpaceChargeCorrection.h:322:50: error: field reference 'fInterpolationSafetyMargin' does not match naming rule [aliceO2-member-name]
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12648/29e513b5f284a9986d6fab4ed23618fc4200bb55/build_O2_fullCI/pretty.html).

ktf committed:
DPL Analysis: introduce PersistentColumn concept

This way the compiler does not need to instanciate the conjuction type
and it will point to exactly which column is not persistent, when needed.

aalkin commented:
Concept is certainly better than a conjunction, however this protection is redundant - these functions are only used internally by `Table` and only with persistent columns pack. If somehow they would be called with a non-persistent pack, the code would simply fail at runtime.

alibuild commented:
Error while checking build/O2/fullCI for ec23d755b304618b35f3c48fe9094039ab2b9cfd at 2024-02-13 02:32:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2-full-system-test-latest/log
task timeout reached .. killing all processes


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
++ GRERR=1
++ [[ 1 == 0 ]]
++ mkdir -p /sw/INSTALLROOT/fa91774b9696f39fd1899ac986ff2828ee0f0b01/slc8_x86-64/o2checkcode/1.0-local62/etc/modulefiles
++ cat
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12648/ec23d755b304618b35f3c48fe9094039ab2b9cfd/build_O2_fullCI/pretty.html).

ktf commented:
Fair enough, closing.
